src/Engines/WebGPU/webgpuShaderProcessorsWGSL.ts

```diff
@@ -0,0 +1,484 @@
+                if (location === undefined) {
+                    Logger.Warn(`Invalid fragment shader: The varying named "${name}" is not declared in the vertex shader! This declaration will be ignored.`);
+                }
+            }
+            else {
+                location = this.webgpuProcessingContext.getVaryingNextLocation(varyingType, this._getArraySize(name, varyingType, preProcessors)[2]);
+                this.webgpuProcessingContext.availableVaryings[name] = location;
+                this._varyingsWGSL.push(`[[location(${location})]] ${name} : ${varyingType};`);
+                this._varyingsDeclWGSL.push(`var<private> ${name} : ${varyingType};`);
+                this._varyingNamesWGSL.push(name);
+            }
+
+            varying = "";
+        }
+        return varying;
+    }
+
+    public attributeProcessor(attribute: string, preProcessors: { [key: string]: string }, processingContext: Nullable<ShaderProcessingContext>) {
+        const attribRegex = new RegExp(/\s*attribute\s+(\S+)\s*:\s*(.+)\s*;/gm);
+        const match = attribRegex.exec(attribute);
+        if (match !== null) {
+            const attributeType = match[2];
+            const name = match[1];
+            const location = this.webgpuProcessingContext.getAttributeNextLocation(attributeType, this._getArraySize(name, attributeType, preProcessors)[2]);
+
+            this.webgpuProcessingContext.availableAttributes[name] = location;
+            this.webgpuProcessingContext.orderedAttributes[location] = name;
+
+            this._attributesWGSL.push(`[[location(${location})]] ${name} : ${attributeType};`);
+            this._attributesDeclWGSL.push(`var<private> ${name} : ${attributeType};`);
+            this._attributeNamesWGSL.push(name);
+            attribute = "";
+        }
+        return attribute;
+    }
+
+    public uniformProcessor(uniform: string, isFragment: boolean, preProcessors: { [key: string]: string }, processingContext: Nullable<ShaderProcessingContext>): string {
+        const match = this.uniformRegexp.exec(uniform);
+        if (match !== null) {
+            let uniformType = match[2];
+            let name = match[1];
+
+            this._addUniformToLeftOverUBO(name, uniformType, preProcessors);
+
+            uniform = "";
+        }
+        return uniform;
+    }
+
+    public textureProcessor(texture: string, isFragment: boolean, preProcessors: { [key: string]: string }, processingContext: Nullable<ShaderProcessingContext>): string {
+        const match = this.textureRegexp.exec(texture);
+        if (match !== null) {
+            const name = match[1]; // name of the variable
+            const type = match[2]; // texture_2d<f32> or array<texture_2d_array<f32>, 5> for eg
+            const isArrayOfTexture = !!match[3];
+            const textureFunc = match[4]; // texture_2d, texture_depth_2d, etc
+            const isStorageTexture = textureFunc.indexOf("storage") > 0;

```

Comment:
> -1 ? (just wondering if it is intentional)

Reply:
Yes, testing for `> 0` is intentional because **storage** can't start at index 0.